Allow to parse dates from long numbers

jiho opened the pull request:
With as.character(), long numbers get converted to scientific
notation which the parse of course fails to recognise. Using
format(..., scientific = FALSE) prevents that.

jiho commented:
Should really have run `check()`. Additional commits fix mistakes.

jiho commented:
Example of what used to fail:

```
> x <- 20140101000000
> as.character(x)
[1] "2.0140101e+13"
> format(x, scientific = FALSE, trim = TRUE)
[1] "20140101000000"
> .num_to_date(x)
[1] "20140101000000"
```

It seems like a bug in `as.character` actually but it is probably safer to fix it here rather than wait for `as.character` to change...

vspinu commented:
Thanks. Could you please add a small test for this?

Could you also squash your commits into one? We try to stick to one-commit-per-change rule to make it easier to navigate the history.

jiho commented:
Regarding performance, `format` would actually be even a bit better:

```
> x <- round(runif(10^7) * 100000)
> system.time(as.character(x))
   user  system elapsed 
  7.786   0.015   7.816 
> system.time(format(x, scientific = FALSE, trim = TRUE))
   user  system elapsed 
  6.412   0.013   6.431 
```

jiho commented:
Tests are added and everything is squashed into one commit

vspinu commented on inst/tests/test-parsers.R:
```diff
@@ -156,6 +156,20 @@ test_that("ymd functions correctly parse dates with no separators and no quotes"
 })
 
+test_that("numbers are correctly converted into character for parsing", {
+  # numbers with 000000 at the end are wrongly converted by as.character
```
They are not wrongly converted. That's the feature which is customized by `scipen` option.

vspinu commented on inst/tests/test-parsers.R:
```diff
@@ -156,6 +156,20 @@ test_that("ymd functions correctly parse dates with no separators and no quotes"
+              equals(as.POSIXct("2010-01-02 00:00:00", tz = "UTC")))
+  expect_that(.num_to_date(20100102000000),
+              equals("20100102000000"))
+  # check for a fix in as.character
```
I don't think we need these tests.  `as.character` is not broken and will never be fixed.

vspinu commented:
This PR made me think, are there other places `as.character` will fail similarly?

jiho commented:
Potential problems:

- `R/parse.r` l. 455 `as.character` is used with randomly generated numbers which may end with many zeros by chance.
- `inst/tests/test-parser.r` l. 520 same as above.
- `R/parse.r` l. 489, `as.character` is used when it is not necessary (`.num_to_date` always returns a `character`)

The other uses in the source code do not seem problematic to me.

jiho commented:
I've addressed the comments above. Let me know what you think. I guess I'll need to squash the commits again but I'll rather do that when all is accepted.

vspinu commented:
> R/parse.r l. 455 as.character is used with randomly generated numbers which may end with many zeros by chance.

Probably not an issue as `as.character` is applied to POSIXct object. `as.character(.POSIXct(1000000000000))` works as expected.

> R/parse.r l. 489, as.character is used when it is not necessary (.num_to_date always returns a character)

Good catch. That makes me think that `.num_to_date` is a misnomer. It doesn't convert to date.

> I've addressed the comments above. Let me know what you think. I guess I'll need to squash the commits again but I'll rather do that when all is accepted.

Yes. Looks good. Please squash. Instead of adding new commits you can always amend previous commit and then force push.

Thanks.

jiho commented:
Squashed commits

vspinu commented:
:+1: Thanks!

jiho commented:
You are welcome. Happy to help.
